felix/dataplane/linux/policy_mgr.go

```diff
@@ -144,7 +122,46 @@ func (m *policyManager) OnUpdate(msg interface{}) {
 	}
 }
 
+func (m *policyManager) cleanUpPolicy(id *proto.PolicyID) {
+	if m.rawEgressOnly {
+		m.updateNeededIPSets(id, nil)
+	}
+	inName := rules.PolicyChainName(rules.PolicyInboundPfx, id)
+	outName := rules.PolicyChainName(rules.PolicyOutboundPfx, id)
+	// As above, we need to clean up in all the tables.
+	m.filterTable.RemoveChainByName(inName)
+	m.filterTable.RemoveChainByName(outName)
+	m.mangleTable.RemoveChainByName(inName)
+	m.mangleTable.RemoveChainByName(outName)
+	m.rawTable.RemoveChainByName(inName)
+	m.rawTable.RemoveChainByName(outName)
+}
+
+func (m *policyManager) updateNeededIPSets(id *proto.PolicyID, neededIPSets set.Set[string]) {
+	if neededIPSets != nil {
+		m.neededIPSets[*id] = neededIPSets
+	} else {
+		delete(m.neededIPSets, *id)
+	}
+	m.ipSetFilterDirty = true
+}
+
 func (m *policyManager) CompleteDeferredWork() error {
-	// Nothing to do, we don't defer any work.
+	if !m.rawEgressOnly {
+		return nil
+	}
+	if !m.ipSetFilterDirty {
+		return nil
+	}
+	m.ipSetFilterDirty = false
+
+	merged := set.New[string]()
+	for _, ipSets := range m.neededIPSets {
+		ipSets.Iter(func(item string) error {
+			merged.Add(item)
+			return nil
+		})
+	}
+	m.ipSetsCallback(merged)
```

Comment:
so in bpf mode, we get here in the first call to CompleteDeferredWork because m.ipSetFilterDirty is true from the start, but after all the merging, there is nothing to program so we do not endup programming anything? While previously things got programmed in OnUpdate rightaway so we endup with say empty sets??

> In BPF mode, we don't want to program linux IP sets unless we're using "untracked" policy, which is partially implemented in iptables. The filter to prevent IP set programming was only set after a policy churn.

I am not sure how it used to be set "after a policy churn"

Reply:
> but after all the merging, there is nothing to program so we do not endup programming anything?

Yes, if there are no do-not-track policies then `neededIPSets` will be empty, so `merged` will be empty and then we'll call ipSetsCallback with that empty set.  That tells the IP sets driver 
not to program any IP sets.

> I am not sure how it used to be set "after a policy churn"

Before, we never called ipSetsCallback unless:

- There was a do-not-track policy (which is rare).
- We received an ActivePolicyRemove message (i.e. _some_ policy was removed from the dataplane).  Then we'd do the same merge process and set an empty filter too.

Reply:
Moving it to the CompleteDeferredWork is just good practice (to avoid recalculating the filter over and over) and it lets me make sure that it always gets done before the first apply.

Reply:
So IPset are added to the `IPSets` in ipsets.go but the policy manager neveer got a chance to say "we actually need only those"? Right? So now we always call it in defered because we start with dirty filter.

Reply:
Yes, exactly.